libopenage/input/action.h

```diff
@@ -17,57 +18,67 @@ struct action_t {
-static constexpr action_t DESELECT = action_t{43, "DESELECT"};
-static constexpr action_t FORWARD = action_t{44, "FORWARD"};
-static constexpr action_t BACK = action_t{45, "BACK"};
-static constexpr action_t PAINT_TERRAIN = action_t{46, "PAINT_TERRAIN"};
+
+static const action_map_t ACTION_MAP = {
+	{"UNDEFINED", action_t{0, "UNDEFINED"}},
+	{"START_GAME", action_t{1, "START_GAME"}},
+	{"STOP_GAME", action_t{2, "STOP_GAME"}},
+	{"TOGGLE_HUD", action_t{3, "TOGGLE_HUD"}},
+	{"SCREENSHOT", action_t{4, "SCREENSHOT"}},
+	{"TOGGLE_DEBUG_OVERLAY", action_t{5, "TOGGLE_DEBUG_OVERLAY"}},
+	{"TOGGLE_DEBUG_GRID", action_t{6, "TOGGLE_DEBUG_GRID"}},
+	{"QUICK_SAVE", action_t{7, "QUICK_SAVE"}},
+	{"QUICK_LOAD", action_t{8, "QUICK_LOAD"}},
+	{"TOGGLE_MODE", action_t{9, "TOGGLE_MODE"}},
+	{"TOGGLE_MENU", action_t{10, "TOGGLE_MENU"}},
+	{"TOGGLE_ITEM", action_t{11, "TOGGLE_ITEM"}},
+	{"TOGGLE_BLENDING", action_t{12, "TOGGLE_BLENDING"}},
+	{"TOGGLE_PROFILER", action_t{13, "TOGGLE_PROFILER"}},
+	{"TOGGLE_CONSTRUCT_MODE", action_t{14, "TOGGLE_CONSTRUCT_MODE"}},
+	{"TOGGLE_UNIT_DEBUG", action_t{15, "TOGGLE_UNIT_DEBUG"}},
+	{"TRAIN_OBJECT", action_t{16, "TRAIN_OBJECT"}},
+	{"ENABLE_BUILDING_PLACEMENT", action_t{17, "ENABLE_BUILDING_PLACEMENT"}},
+	{"DISABLE_SET_ABILITY", action_t{18, "DISABLE_SET_ABILITY"}},
+	{"SET_ABILITY_MOVE", action_t{19, "SET_ABILITY_MOVE"}},
+	{"SET_ABILITY_GATHER", action_t{20, "SET_ABILITY_GATHER"}},
+	{"SET_ABILITY_GARRISON", action_t{21, "SET_ABILITY_GARRISON"}},
+	{"TOGGLE_CONSOLE", action_t{22, "TOGGLE_CONSOLE"}},
+	{"SPAWN_VILLAGER", action_t{23, "SPAWN_VILLAGER"}},
+	{"KILL_UNIT", action_t{24, "KILL_UNIT"}},
+	{"BUILDING_1", action_t{25, "BUILDING_1"}},
+	{"BUILDING_2", action_t{26, "BUILDING_2"}},
+	{"BUILDING_3", action_t{27, "BUILDING_3"}},
+	{"BUILDING_4", action_t{28, "BUILDING_4"}},
+	{"BUILDING_TOWN_CENTER", action_t{29, "BUILDING_TOWN_CENTER"}},
+	{"SWITCH_TO_PLAYER_1", action_t{30, "SWITCH_TO_PLAYER_1"}},
+	{"SWITCH_TO_PLAYER_2", action_t{31, "SWITCH_TO_PLAYER_2"}},
+	{"SWITCH_TO_PLAYER_3", action_t{32, "SWITCH_TO_PLAYER_3"}},
+	{"SWITCH_TO_PLAYER_4", action_t{33, "SWITCH_TO_PLAYER_4"}},
+	{"SWITCH_TO_PLAYER_5", action_t{34, "SWITCH_TO_PLAYER_5"}},
+	{"SWITCH_TO_PLAYER_6", action_t{35, "SWITCH_TO_PLAYER_6"}},
+	{"SWITCH_TO_PLAYER_7", action_t{36, "SWITCH_TO_PLAYER_7"}},
+	{"SWITCH_TO_PLAYER_8", action_t{37, "SWITCH_TO_PLAYER_8"}},
+	{"UP_ARROW", action_t{38, "UP_ARROW"}},
+	{"DOWN_ARROW", action_t{39, "DOWN_ARROW"}},
+	{"LEFT_ARROW", action_t{40, "LEFT_ARROW"}},
+	{"RIGHT_ARROW", action_t{41, "RIGHT_ARROW"}},
+	{"SELECT", action_t{42, "SELECT"}},
+	{"DESELECT", action_t{43, "DESELECT"}},
+	{"FORWARD", action_t{44, "FORWARD"}},
+	{"BACK", action_t{45, "BACK"}},
+	{"PAINT_TERRAIN", action_t{46, "PAINT_TERRAIN"}},
+	{"BUILDING_5", action_t{47, "BUILDING_5"}},
+	{"BUILDING_6", action_t{48, "BUILDING_6"}},
+	{"BUILDING_7", action_t{49, "BUILDING_7"}},
+	{"BUILDING_8", action_t{50, "BUILDING_8"}}
```

Comment:
I know that would mean more work, but having redundancy is bad in almost all cases.
Can't we create a list of "register" those `action_t` objects and register them? This registration could then also happen from around the engine, whereever a new keybinding is created for the appropriate context.